app.ts

```diff
@@ -0,0 +1,1160 @@
+				// is the same of the current session
+				const users = jsonStore.get(store.users)
+
+				user = users.find((u: User) => u.username === decoded.username)
+			} else {
+				// credentials auth
+				const users = jsonStore.get(store.users)
+
+				const username = req.body.username
+				const password = req.body.password
+
+				user = users.find((u: User) => u.username === username)
+
+				if (
+					user &&
+					!(await utils.verifyPsw(password, user.passwordHash))
+				) {
+					user = null
+				}
+			}
+
+			const result = {
+				success: !!user,
+				code: undefined,
+				message: '',
+				user: undefined,
+			}
+
+			if (result.success) {
+				// don't edit the original user object, remove the password from jwt payload
+				const userData: User = Object.assign({}, user)
+				delete userData.passwordHash
+
+				const token = jwt.sign(userData, sessionSecret, {
+					expiresIn: '1d',
+				})
+				userData.token = token
+				req.session.user = userData
+				result.user = userData
+				loginLimiter.resetKey(req.ip)
+			} else {
+				result.code = 3
+				result.message = RESPONSE_CODES['3']
+			}
+
+			res.json(result)
+		} catch (error) {
+			res.json({
+				success: false,
+				message: 'Authentication failed',
+				code: 3,
+			})
+		}
+	}
+)
+
+// logout the user
+app.get('/api/logout', apisLimiter, isAuthenticated, function (req, res) {
+	req.session.destroy((err) => {
+		if (err) {
+			res.json({ success: false, message: err.message })
+		} else {
+			res.json({ success: true, message: 'User logged out' })
+		}
+	})
+})
+
+// update user password
+app.put(
+	'/api/password',
+	apisLimiter,
+	csrfProtection,
+	isAuthenticated,
+	async function (req, res) {
+		try {
+			const users = jsonStore.get(store.users)
+
+			const user = req.session.user
+			const oldUser = users.find((u: { _id: any }) => u._id === user._id)
+
+			if (!oldUser) {
+				return res.json({ success: false, message: 'User not found' })
+			}
+
+			if (
+				!(await utils.verifyPsw(req.body.current, oldUser.passwordHash))
+			) {
+				return res.json({
+					success: false,
+					message: 'Current password is wrong',
+				})
+			}
+
+			if (req.body.new !== req.body.confirmNew) {
+				return res.json({
+					success: false,
+					message: "Passwords doesn't match",
+				})
+			}
+
+			oldUser.passwordHash = await utils.hashPsw(req.body.new)
+
+			req.session.user = oldUser
+
+			await jsonStore.put(store.users, users)
+
+			res.json({
+				success: true,
+				message: 'Password updated',
+				user: oldUser,
+			})
+		} catch (error) {
+			res.json({
+				success: false,
+				message: 'Error while updating passwords',
+				error: error.message,
+			})
+			logger.error('Error while updating password', error)
+		}
+	}
+)
+
+app.get('/health', apisLimiter, function (req, res) {
+	let mqtt: Record<string, any> | boolean
+	let zwave: boolean
+
+	if (gw) {
+		mqtt = gw.mqtt ? gw.mqtt.getStatus() : false
+		zwave = gw.zwave ? gw.zwave.getStatus().status : false
```

Comment:
TypeScript supports [nullish coalescing and optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html). You could turn constructs like these into
```suggestion
		mqtt = gw.mqtt?.getStatus() ?? false
		zwave = gw.zwave?.getStatus().status ?? false
```

Reply:
Yep I didn't refactored it yet :) That's something I really missed on js